File: php/numberIslands.php
<?php

function main()
{
    // Islands array
    $islands = [
        [1, 1, 0, 0, 0, 0, 0],
        [0, 1, 0, 1, 0, 0, 1],
        [0, 0, 1, 1, 0, 0, 1],
        [0, 1, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 1, 1],
        [0, 0, 0, 0, 0, 0, 0],
        [0, 0, 1, 0, 0, 0, 0],
    ];

    // Calculating number rows and columns
    $rows    = count($islands);
    $columns = count($islands[0]);
    $count   = 0;

    // Array of possitions to check
    $positions = function($i, $j) use ($columns, $rows) {
        return [
            [$j != $columns - 1, $i, $j + 1], // Next
            [$j != 0,            $i, $j - 1], // Prev
            [$i != 0,            $i - 1, $j], // Above
            [$i != $rows - 1,    $i + 1, $j], // Below
        ];
    };

    // Iterate each row
    for ($i = 0; $i < $rows; $i++) {

        // Iterate each column
        for ($j = 0; $j < $columns; $j++) {

            // Check for islands
            if ($islands[$i][$j] == 1) {

                // Checking adjacent possitions to find a required value
                foreach ($positions($i, $j) as $key) {
                    if ($key[0] && $islands[$key[1]][$key[2]] == 1) {

                      // If adjacent value found increment count and reset found value (to mark as found)
                      $count++;
                      $islands[$key[1]][$key[2]] = 0;
                      var_dump("x-axis: {$j} -- y-axis: {$i}");
                    }
                }
                $islands[$i][$j] = 0;
            }
        }
    }
    var_dump($count);
}
main();

File: python/numberIslands.py
def grid():
    return [
        [1, 0, 0, 0, 0, 0, 0],
        [0, 1, 0, 1, 0, 0, 0],
        [0, 0, 0, 1, 0, 0, 0],
        [0, 1, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 1, 1],
        [0, 0, 1, 0, 0, 0, 0],
        [0, 0, 1, 1, 0, 0, 0],
    ]

def countIslands(grid):
    count    = 0
    rows     = len(grid)
    columns  = len(grid[0])
    used     = [[False for j in xrange(columns)] for i in xrange(rows)]

    for x in (xrange(rows)):
        for y in (xrange(columns)):
            if grid[x][y] == 1 and not used[x][y]:
                checkNeighbours(grid, used, rows, columns, x, y)
                count += 1
    return count

def checkNeighbours(grid, used, rows, columns, x, y):
    if grid[x][y] == 0 or used[x][y]:
        return
    used[x][y] = True

    if x != 0:
        checkNeighbours(grid, used, rows, columns, x - 1, y)
    if x != rows - 1:
        checkNeighbours(grid, used, rows, columns, x + 1, y)
    if y != 0:
        checkNeighbours(grid, used, rows, columns, x, y - 1)
    if y != columns -1:
        checkNeighbours(grid, used, rows, columns, x, y + 1)

print countIslands(grid())
